src/Pos/Block/Logic/Slog.hs

```diff
@@ -171,7 +171,11 @@ slogApplyBlocks blunds = do
             SomeBatchOp $
             GS.PutTip $ headerHash $ NE.last $ getOldestFirst blunds
     sanityCheckDB
-    putSlottingData =<< GS.getSlottingData
+
+    slottingData <- GS.getAllSlottingData
+    -- TODO: Could this be more efficient?
```

Comment:
Yes, it can. I think it's ok to rely on the fact that slotting data changes only when we apply genesis block. Also there is invariant that all applied blocks in this function have same epoch (unfortunately, it's not written here, but it's written in `Pos.Block.Logic.Internal` which calls this function).
So we can't apply more than one genesis block and if we apply genesis block it must be head.
The simplest solution is:
• if we apply genesis block for the `i`-th epoch, then we should use `putEpochSlottingData` for the `i + 1`-th epoch;
• otherwise don't do anything.

Apart from that I just noticed that `slogRollbackBlocks` doesn't update in-memory slotting data, though it can be changed too. It will be great if you fix it too. But do it only if you want, it's not your mistake.